scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/IncrementalSimulationDriver.java

```diff
@@ -0,0 +1,182 @@
+package gov.nasa.jpl.aerie.scheduler;
+
+import gov.nasa.jpl.aerie.merlin.driver.ActivityInstanceId;
+import gov.nasa.jpl.aerie.merlin.driver.MissionModel;
+import gov.nasa.jpl.aerie.merlin.driver.SerializedActivity;
+import gov.nasa.jpl.aerie.merlin.driver.SimulationResults;
+import gov.nasa.jpl.aerie.merlin.driver.engine.SimulationEngine;
+import gov.nasa.jpl.aerie.merlin.driver.engine.TaskId;
+import gov.nasa.jpl.aerie.merlin.driver.timeline.LiveCells;
+import gov.nasa.jpl.aerie.merlin.driver.timeline.TemporalEventSource;
+import gov.nasa.jpl.aerie.merlin.protocol.types.Duration;
+import org.apache.commons.lang3.tuple.Pair;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class IncrementalSimulationDriver {
+
+  private Duration curTime;
+  private SimulationEngine engine;
+  private LiveCells cells;
+  private TemporalEventSource timeline;
+  private final MissionModel<?> missionModel;
+
+  //mapping each activity name to its task id (in String form) in the simulation engine
+  private Map<ActivityInstanceId, TaskId> plannedDirectiveToTask;
+  //and the opposite
+  private Map<TaskId, ActivityInstanceId> taskToPlannedDirective;
+
+  //simulation results so far
+  private SimulationResults lastSimResults;
+  //cached simulation results cover the period [Duration.ZERO, lastSimResultsEnd]
+  private Duration lastSimResultsEnd = Duration.ZERO;
+
+  //List of activities simulated since the last reset
+  private final List<SimulatedActivity> activitiesInserted = new ArrayList<>();
+
+  record SimulatedActivity(Duration start, SerializedActivity activity, ActivityInstanceId id) {}
+
+  public IncrementalSimulationDriver(MissionModel<?> missionModel){
+    this.missionModel = missionModel;
+    initSimulation();
+  }
+
+  private void initSimulation(){
+    plannedDirectiveToTask = new HashMap<>();
+    taskToPlannedDirective = new HashMap<>();
```

Comment:
It's good that these member variables are being assigned a value at construction time but without setting these directly within the class member variable declaration (lines 31 & 33 above) or within the class constructor we cannot guarantee that these fields are assigned and not null. Could you make these member variables final and either merge this assignment with their declarations or move them to the constructor? Thanks :)

Reply:
Hmm, `initSimulation` can get called multiple times over the lifetime of an `IncrementalSimulationDriver`.

Maybe we could initialize the hashmaps in the constructor, then call `.clear()` in `initSimulation`?

Reply:
Yeah, that'd make sense. Or some sort of builder pattern to rein in and consolidate the mutable member variables.